src/cpp/ext/otel/otel_plugin.h

```diff
@@ -217,53 +186,233 @@ class OpenTelemetryPluginBuilderImpl {
+  void SetGauge(
+      grpc_core::GlobalInstrumentsRegistry::GlobalDoubleGaugeHandle handle,
+      double value, absl::Span<const absl::string_view> label_values,
+      absl::Span<const absl::string_view> optional_values) override {}
+  // TODO(yashkt, yijiem): implement async instrument.
+  void AddCallback(grpc_core::RegisteredMetricCallback* callback) override {}
+  void RemoveCallback(grpc_core::RegisteredMetricCallback* callback) override {}
+  grpc_core::ClientCallTracer* GetClientCallTracer(
+      const grpc_core::Slice& path, bool registered_method,
+      std::shared_ptr<grpc_core::StatsPlugin::ScopeConfig> scope_config)
+      override;
+  grpc_core::ServerCallTracer* GetServerCallTracer(
+      std::shared_ptr<grpc_core::StatsPlugin::ScopeConfig> scope_config)
+      override;
+
+  const absl::AnyInvocable<bool(const grpc_core::ChannelArgs& /*args*/) const>&
+  server_selector() const {
+    return server_selector_;
+  }
+  const absl::AnyInvocable<bool(absl::string_view /*target*/) const>&
+  target_attribute_filter() const {
+    return target_attribute_filter_;
+  }
+  const absl::AnyInvocable<bool(absl::string_view /*generic_method*/) const>&
+  generic_method_attribute_filter() const {
+    return generic_method_attribute_filter_;
+  }
+  const std::vector<std::unique_ptr<InternalOpenTelemetryPluginOption>>&
+  plugin_options() const {
+    return plugin_options_;
   }
+  Client& client() { return client_; }
+  Server& server() { return server_; }
 
-  std::bitset<64> active_mask_;
+  // Instruments for per-call metrics.
+  Client client_;
+  Server server_;
+  // Instruments for non-per-call metrics.
+  struct Disabled {};
+  using InstrumentType = absl::variant<
+      Disabled, std::unique_ptr<opentelemetry::metrics::Counter<uint64_t>>,
+      std::unique_ptr<opentelemetry::metrics::Counter<double>>,
+      std::unique_ptr<opentelemetry::metrics::Histogram<uint64_t>>,
+      std::unique_ptr<opentelemetry::metrics::Histogram<double>>>;
+  struct Instrument {
+    InstrumentType otel_instrument;
+    // Stores label keys and optional label keys for each instrument at
+    // construction time.
+    std::vector<absl::string_view> label_keys;
+    std::vector<absl::string_view> optional_label_keys;
+  };
+  std::vector<Instrument> instruments_;
+  opentelemetry::nostd::shared_ptr<opentelemetry::metrics::MeterProvider>
+      meter_provider_;
+  absl::AnyInvocable<bool(absl::string_view /*target*/) const> target_selector_;
+  absl::AnyInvocable<bool(const grpc_core::ChannelArgs& /*args*/) const>
+      server_selector_;
+  absl::AnyInvocable<bool(absl::string_view /*target*/) const>
+      target_attribute_filter_;
+  absl::AnyInvocable<bool(absl::string_view /*generic_method*/) const>
+      generic_method_attribute_filter_;
+  std::vector<std::unique_ptr<InternalOpenTelemetryPluginOption>>
+      plugin_options_;
+  std::shared_ptr<std::set<absl::string_view>> optional_label_keys_;
```

Comment:
Let's not save the optional label keys. Instead, let's save a bit set along with each created instrument to denote whether an optional label in a particular position needs to be recorded or not

Reply:
`std::bitset<N>` is a fixed-size sequence of N bits (N needs to be specified during compile-time), are we certain that the optional label keys size for an instrument will be less than a reasonable bound (say 64)? Or we actually want to be dynamic and store something like a `std::vector<std::bitset<64>>` with each instrument?

Reply:
I'm fine with saying that there can be no more than 64 optional labels for a given metric.

We do need to store a separate bitset for each instrument, because the set and order of optional labels will be different for each one.  I think we want something like this:

```
struct InstrumentData {
  Instrument instrument;
  std::bitset<64> optional_labels_bits;
};
std::vector<InstrumentData> instruments_;
```

Reply:
agreed